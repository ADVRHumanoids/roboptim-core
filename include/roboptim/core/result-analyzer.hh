// Copyright (C) 2015 by Benjamin Chr√©tien, CNRS-UM LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_RESULT_ANALYZER_HH
# define ROBOPTIM_CORE_RESULT_ANALYZER_HH

# include <roboptim/core/problem.hh>
# include <roboptim/core/differentiable-function.hh>

namespace roboptim
{
  template <typename T>
  class ResultAnalyzer
  {
  public:
    typedef GenericFunctionTraits<T>              functionTraits_t;
    typedef Problem<T>                            problem_t;
    typedef Result                                optimResult_t;
    typedef typename functionTraits_t::argument_t argument_t;
    typedef typename functionTraits_t::result_t   result_t;
    typedef typename functionTraits_t::gradient_t gradient_t;
    typedef typename functionTraits_t::jacobian_t jacobian_t;
    typedef typename problem_t::interval_t        interval_t;
    typedef typename problem_t::size_type         size_type;
    typedef typename problem_t::value_type        value_type;

    typedef GenericDifferentiableFunction<T> differentiableFunction_t;

    ResultAnalyzer (const problem_t& pb, const optimResult_t& res);
    virtual ~ResultAnalyzer ();

    struct QueryData
    {
      operator bool () const;

      virtual bool isValid () const = 0;
    };

    struct LICQData : public QueryData
    {
      LICQData ();
      bool isValid () const;

      size_type rank;
      size_type max_rank;
    };

    struct KKTData : public QueryData
    {
      KKTData ();
      bool isValid () const;

      gradient_t grad_L;
      value_type eps;
    };

    struct NullGradientData : public QueryData
    {
      NullGradientData();
      bool isValid () const;

      size_type null_rows;
    };

    /// \brief Check the LICQ conditions.
    LICQData checkLICQ () const;

    /// \brief Check the KKT conditions.
    KKTData checkKKT () const;

    /// \brief Check for null gradients.
    NullGradientData checkNullGradient () const;

  private:
    void computeJacobian () const;
    void computeActiveJacobian () const;

    size_type computeRank (const jacobian_t& jac) const;

    bool hasLambda () const;

  private:
    /// \brief Problem.
    const problem_t pb_;

    /// \brief Optimization result.
    const optimResult_t res_;

    /// \brief Jacobian of the active constraints.
    mutable jacobian_t activeJac_;

    /// \brief Jacobian of the constraints.
    mutable jacobian_t jac_;

    /// \brief Epsilon.
    value_type eps_;
  };
} // end of namespace roboptim

# include <roboptim/core/result-analyzer.hxx>

#endif //! ROBOPTIM_CORE_RESULT_ANALYZER_HH
