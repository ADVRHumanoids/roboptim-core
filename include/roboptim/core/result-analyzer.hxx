// Copyright (C) 2015 by Benjamin Chrétien, CNRS-UM LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_RESULT_ANALYZER_HXX
# define ROBOPTIM_CORE_RESULT_ANALYZER_HXX

# include <vector>
# include <limits>

# include <Eigen/OrderingMethods>

# include <roboptim/core/indent.hh>
# include <roboptim/core/util.hh>

namespace roboptim
{
  template <typename T>
  ResultAnalyzer<T>::ResultAnalyzer (const problem_t& pb,
                                     const optimResult_t& res)
  : pb_ (pb),
    res_ (res),
    activeJac_ (),
    jac_ (),
    activeCstrIndices_ (),
    eps_ (1e-12)
  {
  }

  template <typename T>
  ResultAnalyzer<T>::~ResultAnalyzer ()
  {
  }

  template <typename T>
  bool ResultAnalyzer<T>::hasLambda () const
  {
    size_type n = pb_.function ().inputSize ();
    size_type m = pb_.constraintsOutputSize ();

    return (res_.lambda.size () == n + m + 1);
  }

  template <typename T>
  bool ResultAnalyzer<T>::QueryData::operator_bool () const
  {
    return isValid ();
  }

  template <typename T>
  ResultAnalyzer<T>::LICQData::LICQData ()
  : rank (0),
    max_rank (0)
  {
  }

  template <typename T>
  bool ResultAnalyzer<T>::LICQData::isValid () const
  {
    return rank == max_rank;
  }

  template <typename T>
  std::ostream& ResultAnalyzer<T>::LICQData::print (std::ostream& o) const
  {
    o << "LICQ conditions: ";
    if (!isValid ()) o << "not ";
    o << "satisfied";

    o << incindent << iendl << "rank: " << rank
      << iendl << "max_rank: " << max_rank << decindent;

    return o;
  }

  template <typename T>
  ResultAnalyzer<T>::KKTData::KKTData ()
  : grad_L (),
    lambda (),
    eps (std::numeric_limits<value_type>::epsilon ())
  {
  }

  template <typename T>
  bool ResultAnalyzer<T>::KKTData::isValid () const
  {
    return grad_L.size () > 0
      && grad_L.norm () < eps
      && (lambda.array () >= 0.).all ();
  }

  template <typename T>
  std::ostream& ResultAnalyzer<T>::KKTData::print (std::ostream& o) const
  {
    o << "KKT conditions: ";
    if (isValid ()) o << "satisfied";
    else o << "not satisfied";

    o << incindent;

    if (grad_L.size () == 0)
      o << iendl << "∇L(x*,λ*) could not be computed";
    else
      o << iendl << "∇L(x*,λ*) = " << grad_L;

    if (lambda.size () == 0)
      o << iendl << "λ was not provided";
    else
      o << iendl << "λ = " << lambda;

    o << decindent;

    return o;
  }

  template <typename T>
  ResultAnalyzer<T>::NullGradientData::NullGradientData ()
  : null_rows (0),
    constraint_indices ()
  {
  }

  template <typename T>
  bool ResultAnalyzer<T>::NullGradientData::isValid () const
  {
    return null_rows == 0;
  }

  template <typename T>
  std::ostream& ResultAnalyzer<T>::NullGradientData::print (std::ostream& o) const
  {
    o << "Null gradient condition:";
    if (isValid ())
    {
      return o << " satisfied";
    }

    o << " not satisfied" << incindent;
    for (typename std::map<const constraint_t, std::vector<size_type> >::const_iterator
         c  = constraint_indices.begin ();
         c != constraint_indices.end (); ++c)
    {
      o << iendl << "- Constraint: " << incindent
        << *(c->first) << decindent
        << iendl << "  Null gradient indices: " << c->second;
    }
    o << decindent;

    return o;
  }

  template <typename T>
  void ResultAnalyzer<T>::computeJacobian () const
  {
    if (jac_.cols () + jac_.rows () != 0)
      return;

    jac_ = pb_.jacobian (res_.x);
  }

  template <typename T>
  void ResultAnalyzer<T>::computeActiveJacobian () const
  {
    if (activeJac_.cols () + activeJac_.rows ()!= 0)
      return;

    size_type n = pb_.function ().inputSize ();
    size_type m = pb_.constraintsOutputSize ();

    const argument_t& x = res_.x;

    // 1) Argument bounds
    std::vector<size_type> activeBounds;
    for (size_type i = 0; i < n; ++i)
    {
      const interval_t& bounds = pb_.argumentBounds ()[static_cast<size_t> (i)];
      if (std::abs (x[i] - bounds.first) < eps_
          || std::abs (x[i] - bounds.second) < eps_)
        activeBounds.push_back (i);
    }

    // 2) Constraints
    std::vector<size_type> activeConstraints;
    size_type idx = 0;
    size_t cstr_idx = 0;

    for (typename problem_t::constraints_t::const_iterator
         cstr  = pb_.constraints ().begin ();
         cstr != pb_.constraints ().end (); ++cstr, ++cstr_idx)
    {
      const typename problem_t::intervals_t&
        bounds = pb_.boundsVector ()[cstr_idx];
      result_t val = (**cstr) (x);
      for (size_type j = 0; j < val.size (); ++j, ++idx)
      {
        size_t jj = static_cast<size_t> (j);
        if (std::abs (val[j] - bounds[jj].first) < eps_
            || std::abs (val[j] - bounds[jj].second) < eps_)
        {
          activeConstraints.push_back (idx);

          ConstraintIndex index;
          index.global = cstr_idx;
          index.local = j;
          index.active = static_cast<size_type> (activeCstrIndices_.size ());
          activeCstrIndices_.push_back (index);
        }
      }
    }

    size_type m_act = static_cast<size_type> (activeBounds.size ()
        + activeConstraints.size ());
    activeJac_.resize (m_act, n);
    activeJac_.setZero ();

    for (size_type i = 0; i < static_cast<size_type> (activeBounds.size ()); ++i)
    {
      // FIXME: sparse
      activeJac_ (i, activeBounds[static_cast<size_t> (i)]) = 1.;
    }

    computeJacobian ();

    for (size_t i = 0; i < activeConstraints.size (); ++i)
    {
      // Index in the global Jacobian of active constraints
      size_type idx = static_cast<size_type> (activeBounds.size () + i);
      // FIXME: sparse
      activeJac_.row (idx) = jac_.row (activeConstraints[i]);
    }
  }

  template <>
  inline typename ResultAnalyzer<EigenMatrixSparse>::size_type
  ResultAnalyzer<EigenMatrixSparse>::computeRank (const jacobian_t& jac) const
  {
    Eigen::SparseQR<jacobian_t, Eigen::COLAMDOrdering<int> > qr (jac);
    return static_cast<size_type> (qr.rank ());
  }

  template <typename T>
  typename ResultAnalyzer<T>::size_type
  ResultAnalyzer<T>::computeRank (const jacobian_t& jac) const
  {
    Eigen::FullPivLU<jacobian_t> lu (jac);
    return static_cast<size_type> (lu.rank ());
  }

  template <typename T>
  typename ResultAnalyzer<T>::LICQData
  ResultAnalyzer<T>::checkLICQ () const
  {
    computeActiveJacobian ();

    LICQData licq;

    licq.max_rank = static_cast<size_type> (std::min (activeJac_.rows (),
                                                      activeJac_.cols ()));
    licq.rank = computeRank (activeJac_);

    return licq;
  }

  template <typename T>
  typename ResultAnalyzer<T>::KKTData
  ResultAnalyzer<T>::checkKKT () const
  {
    computeActiveJacobian ();

    KKTData kkt;

    if (!hasLambda ())
      return kkt;

    size_type n = pb_.function ().inputSize ();
    size_type m = pb_.constraintsOutputSize ();
    const argument_t& x = res_.x;

    const differentiableFunction_t* diffObj
      = pb_.function ().template castInto<differentiableFunction_t> ();
    kkt.grad_L = diffObj->gradient (x);

    // Argument bounds
    kkt.grad_L += res_.lambda.segment (0, n);

    // Constraints
    kkt.grad_L += res_.lambda.segment (n, m) * jac_;

    // Store lambda
    kkt.lambda = res_.lambda;

    return kkt;
  }

  template <typename T>
  typename ResultAnalyzer<T>::NullGradientData
  ResultAnalyzer<T>::checkNullGradient () const
  {
    computeActiveJacobian ();

    NullGradientData null_grad;

    for (typename std::vector<ConstraintIndex>::const_iterator
         ci  = activeCstrIndices_.begin ();
         ci != activeCstrIndices_.end (); ++ci)
    {
      size_t g_idx = ci->global;
      size_type l_idx = ci->local;
      size_type as_idx = ci->active;

      if (activeJac_.row (as_idx).norm () < eps_)
      {
        null_grad.null_rows++;
        null_grad.constraint_indices[pb_.constraints ()[g_idx]].push_back (l_idx);
      }
    }

    return null_grad;
  }

  template <typename T>
  std::ostream&
  operator<< (std::ostream& o,
              const typename ResultAnalyzer<T>::LICQData& d)
  {
    return d.print (o);
  }

  template <typename T>
  std::ostream&
  operator<< (std::ostream& o,
              const typename ResultAnalyzer<T>::KKTData& d)
  {
    return d.print (o);
  }

  template <typename T>
  std::ostream&
  operator<< (std::ostream& o,
              const typename ResultAnalyzer<T>::NullGradientData& d)
  {
    return d.print (o);
  }
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_RESULT_ANALYZER_HXX
