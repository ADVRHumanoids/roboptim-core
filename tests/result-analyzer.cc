// Copyright (C) 2015 by Benjamin Chrétien, CNRS-UM LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <boost/make_shared.hpp>

#include <roboptim/core/io.hh>
#include <roboptim/core/result-analyzer.hh>

using namespace roboptim;

typedef boost::mpl::list< ::roboptim::EigenMatrixDense,
                          ::roboptim::EigenMatrixSparse> functionTypes_t;

boost::shared_ptr<boost::test_tools::output_test_stream> output;

// Define f(x) = x * x
template <typename T>
struct Square : public GenericDifferentiableFunction<T>
{
  ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
    (GenericDifferentiableFunction<T>);

  explicit Square (size_type n)
    : GenericDifferentiableFunction<T> (n, 1, "x * x")
  {
  }

  void impl_compute (result_ref res,
                     const_argument_ref x) const
  {
    res[0] = x.dot (x);
  }

  void impl_gradient (gradient_ref grad,
                      const_argument_ref x,
                      size_type = 0) const
  {
    grad = 2. * x;
  }
};

template <>
void Square<EigenMatrixSparse>::impl_gradient
(gradient_ref grad, const_argument_ref x, size_type) const
{
  grad = x.sparseView ();
  grad *= 2.;
}


template <typename T>
class NullSolver : public Solver<T>
{
public:
  typedef Solver<T> parent_t;
  typedef typename parent_t::problem_t problem_t;
  typedef typename problem_t::function_t::argument_t argument_t;

  NullSolver (const problem_t& pb)
    : parent_t (pb)
  {
  }

  ~NullSolver ()
  {
  }

  void solve ()
  {
    (*output) << "solve ()" << std::endl;

    SolverError error ("the null solver always fails.");
    error.lastState () = Result (1, 1);

    parent_t::result_ = error;
  }
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE_TEMPLATE (result_analyzer, T, functionTypes_t)
{
  output = retrievePattern ("result-analyzer");

  typedef NullSolver<T> solver_t;
  typedef typename solver_t::problem_t problem_t;

  typedef typename problem_t::function_t      function_t;
  typedef typename function_t::argument_t     argument_t;
  typedef typename problem_t::intervals_t     intervals_t;
  typedef typename problem_t::scaling_t       scaling_t;

  typedef Square<T>                           square_t;

  // For shared_ptr constructor
  boost::shared_ptr<square_t>
    f = boost::make_shared<square_t> (2);

  problem_t pb (f);
  argument_t x (2);
  x << 0., 0.;
  pb.startingPoint () = x;

  typename square_t::names_t names (2);
  names[0] = "x₀";
  names[1] = "x₁";
  pb.argumentNames () = names;

  for (std::size_t i = 0; i < 2; ++i)
    pb.argumentBounds ()[i] = function_t::makeInterval (-1., 3.);

  boost::shared_ptr<square_t>
    cstr = boost::make_shared<square_t>  (2);
  intervals_t intervals (1);
  scaling_t scaling (1, 1);
  for (size_t i = 0; i < intervals.size (); ++i)
    intervals[i] = Function::makeInterval (0., 9.);
  pb.addConstraint (cstr, intervals, scaling);

  // A = {g}, but ∂g = 0
  {
    Result res (pb.function ().inputSize (), pb.constraintsOutputSize ());
    res.x = (argument_t (2) << 0., 0.).finished ();
    res.lambda = (argument_t (4) << 0., 0., 0., 1.).finished ();

    ResultAnalyzer<T> ra (pb, res);

    (*output) << "x = " << res.x << std::endl;
    (*output) << "λ = " << res.lambda << std::endl;

    typename ResultAnalyzer<T>::LICQData licq = ra.checkLICQ ();
    BOOST_CHECK (!licq);
    (*output) << licq << std::endl;

    typename ResultAnalyzer<T>::KKTData kkt = ra.checkKKT ();
    BOOST_CHECK (kkt);
    (*output) << kkt << std::endl;

    typename ResultAnalyzer<T>::NullGradientData ngd = ra.checkNullGradient ();
    BOOST_CHECK (!ngd);
    (*output) << ngd << std::endl;
  }

  // A = {x,g}, but gradients linearly dependent
  {
    Result res (pb.function ().inputSize (), pb.constraintsOutputSize ());
    res.x = (argument_t (2) << 0., 3.).finished ();
    res.lambda = (argument_t (4) << 1., 1., -1., 1.).finished ();

    ResultAnalyzer<T> ra (pb, res);

    (*output) << "x = " << res.x << std::endl;
    (*output) << "λ = " << res.lambda << std::endl;

    typename ResultAnalyzer<T>::LICQData licq = ra.checkLICQ ();
    BOOST_CHECK (!licq);
    (*output) << licq << std::endl;

    typename ResultAnalyzer<T>::KKTData kkt = ra.checkKKT ();
    BOOST_CHECK (!kkt);
    (*output) << kkt << std::endl;

    typename ResultAnalyzer<T>::NullGradientData ngd = ra.checkNullGradient ();
    BOOST_CHECK (ngd);
    (*output) << ngd << std::endl;
  }

  // TODO: A = {g}

  // A = {x}
  {
    Result res (pb.function ().inputSize (), pb.constraintsOutputSize ());
    res.x = (argument_t (2) << -1., 1.).finished ();
    res.lambda = (argument_t (4) << 1., 1., -1., 1.).finished ();

    ResultAnalyzer<T> ra (pb, res);

    (*output) << "x = " << res.x << std::endl;
    (*output) << "λ = " << res.lambda << std::endl;

    typename ResultAnalyzer<T>::LICQData licq = ra.checkLICQ ();
    BOOST_CHECK (licq);
    (*output) << licq << std::endl;

    typename ResultAnalyzer<T>::KKTData kkt = ra.checkKKT ();
    BOOST_CHECK (!kkt);
    (*output) << kkt << std::endl;

    typename ResultAnalyzer<T>::NullGradientData ngd = ra.checkNullGradient ();
    BOOST_CHECK (ngd);
    (*output) << ngd << std::endl;
  }

  (*output) << pb << std::endl;

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
