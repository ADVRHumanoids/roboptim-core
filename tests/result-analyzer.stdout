LICQ conditions: not satisfied
KKT conditions: not satisfied
Null gradient condition: not satisfied
  - Constraint: x * x (differentiable function)
    Null gradient indices: 0
LICQ conditions: not satisfied
KKT conditions: not satisfied
Null gradient condition: satisfied
LICQ conditions: satisfied
KKT conditions: not satisfied
Null gradient condition: satisfied
Problem:
  x * x (differentiable function)
  Objective scaling: 1
  Arguments bounds: (-1, 3), (-1, 3)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  Number of constraints: 1
  Constraint 0
      x * x (differentiable function)
      Bounds: (0, 9)
      Scaling: 1
      Initial value: [1](0)
  Starting point: [2](0,0)
  Starting value: [1](0)
  Infinity value (for all functions): inf
